interp/src/flatten/structures/index_trait.rs

```diff
@@ -0,0 +1,203 @@
+// TODO(griffin): Replace with cranelift_entity if this ends up being the same
+pub trait IndexRef: Copy + Eq {
+    fn index(&self) -> usize;
+    fn new(input: usize) -> Self;
+}
+
+/// This macro is used to implement the IndexRef trait for a type that wraps an
+/// unsigned integer value. By default, the macro will implement the trait using
+/// a [`u32`](std::u32) as the backing type. However, if a different backing type
+/// is desired, it can be specified as the second argument.
+macro_rules! impl_index {
```

Comment:
Potential alternative you might enjoy: https://github.com/cucapra/filament/blob/main/src/utils/idx.rs

Reply:
Yeah I honestly cannot decide which I prefer. There are certainly some merits to either approach. The main issue I see for the interpreter is that we want many types of indices which themselves are not necessarily "representing" a singular type in the original IR. For example we want local port indices for definitions which themselves can be turned into global indices for the state table and both philosophically represent ports from a certain view but are not the `IR::Port` type. And it seems awkward to invent new empty types to associate with `Idx` for that purpose. Hopefully that makes some sort of sense?